\chapter{Use of ontologies}
\label{chap:implemented-ontologies}

  Throughout \texttt{sparqling-genomics} we use ontologies defined by external
  working groups to increase interoperability with foreign systems.  In this
  chapter, we explain which ontologies we use, and how we use them.

\section{Dublin Core Terms}

  In the web interface, we use the Dublin Core Terms ontology \citep{dcmi-terms}
  to define organizations, collections, datasets, and samples.  Table
  \ref{table:dcterms-usage} provides an overview of the properties we use.

  \hypersetup{urlcolor=black}
  \begin{table}[H]
    \begin{tabularx}{\textwidth}{*{2}{!{\VRule[-1pt]}l}!{\VRule[-1pt]}X}
      \headrow
      \textbf{Term} & \textbf{Used as} & \textbf{Description}\\
      \evenrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#terms-Agent}
           {\texttt{dcterms:Agent}}
      & Class
      & Used by the web interface to describe the organization that produced
      a collection or dataset.\\
      \oddrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#dcmitype-Collection}
           {\texttt{dctype:Collection}}
      & Class
      & Used by the portal page in the web interface as a filter mechanism to browse
        \texttt{dctype:Dataset}s..\\
      \evenrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#dcmitype-Dataset}
           {\texttt{dctype:Dataset}}
      & Class
      & Used by the portal page in the web interface to describe data in a graph.\\
      \oddrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#terms-isPartOf}
           {\texttt{dcterms:isPartOf}}
      & Predicate
      & Used by the portal page in the web interface to express that a \texttt{dctype:Dataset}
        is linked to a \texttt{dctype:Collection}.\\
      \evenrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#elements-title}
           {\texttt{dcterms:title}}
      & Predicate
      & Used by the web interface to name a \texttt{dctype:Collection} or a \texttt{dctype:Dataset}.\\
      \oddrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#elements-publisher}
           {\texttt{dcterms:publisher}}
      & Predicate
      & Used by the web interface to identify the organization that published
      a collection or dataset.\\
      \evenrow
      \href{http://www.dublincore.org/specifications/dublin-core/dcmi-terms/2012-06-14/\#elements-description}
           {\texttt{dc:description}}
      & Predicate
      & Used by the web interface to provide a textual description of a
        \texttt{dctype:Collection} or a \texttt{dctype:Dataset}.\\
    \end{tabularx}
    \caption{\small Terms used from the Dublic Core Terms ontology.}
    \label{table:dcterms-usage}
  \end{table}
  \hypersetup{urlcolor=LinkGray}

\section{Describing genomic positions with FALDO}

  When describing the position of a nucleotide relative to its reference
  genome, we use the Feature Annotation Location Description Ontology (FALDO)
  \citep{Bolleman2016}. Table \ref{table:faldo-usage} provides an overview of
  the properties we use.

  \hypersetup{urlcolor=black}
  \begin{table}[H]
    \begin{tabularx}{\textwidth}{*{2}{!{\VRule[-1pt]}l}!{\VRule[-1pt]}X}
      \headrow
      \textbf{Term} & \textbf{Used as} & \textbf{Usage}\\
      \evenrow
      \href{http://biohackathon.org/resource/faldo\#position}{\texttt{faldo:position}}
      & Predicate
      & Used by \texttt{vcf2rdf} to describe the basepair position within a
      chromosome or contig.\\
      \oddrow
      \href{http://biohackathon.org/resource/faldo\#reference}{\texttt{faldo:reference}}
      & Predicate
      & Used by \texttt{vcf2rdf} to describe the chromosome or contig to which
      the \texttt{faldo:position} is relative to.\\
    \end{tabularx}
    \caption{\small Terms used from FALDO.}
    \label{table:faldo-usage}
  \end{table}
  \hypersetup{urlcolor=LinkGray}

\section{Custom terms}

  Sometimes we miss the right term to describe a statement.  In such cases we
  decide on a new term that is then part of the \emph{SPARQLing-genomics
    ontology}.  The use of these terms is subject to change in upcoming versions
  of \texttt{sparqling-genomics}.  Table \ref{table:sg-usage} summarizes the
  terms that are waiting to be replaced by an external ontology.

  \hypersetup{urlcolor=black}
  \begin{table}[H]
    \begin{tabularx}{\textwidth}{*{2}{!{\VRule[-1pt]}l}!{\VRule[-1pt]}X}
      \headrow
      \textbf{Term} & \textbf{Used as} & \textbf{Usage}\\
      \evenrow
      \href{http://sparqling-genomics/Origin}{\texttt{sg:Origin}}
      & Class
      & Used by the command-line tools (see chapter \ref{chap:command-line}
           {\color{LinkGray}`\nameref{chap:command-line}'})
      to point to the file or resource from which information was derived.\\
      \oddrow
      \texttt{sg:containsDataFor}
      & Class
      & Used by the portal page in the web interface to express which graph
      (object) belongs to which dataset (subject).\\
      \evenrow
      \texttt{sg:mappedToGenomeAssembly}
      & Class
      & Used by the portal page in the web interface to describe the reference
      genome assembly used for a \texttt{dctype:Dataset}.\\
      \oddrow
      \texttt{sg:Sample}
      & Class
      & Used by \texttt{vcf2rdf} to describe a sample.\\
    \end{tabularx}
    \caption{\small Terms made up by us.}
    \label{table:sg-usage}
  \end{table}
  \hypersetup{urlcolor=LinkGray}

