\chapter{Web interface}
\label{chap:web-interface}
\index{sg-web}

  The command-line programs described in chapter \refer{chap:command-line}
  provide data for layer \t{0}.  With the web interface described in
  this chapter, this data can be further structured, analyzed, and shared.
  In short, with the web interface you can:
  \begin{itemize}
  \item Manage data and data access across RDF stores;
  \item Write, execute, and share SPARQL queries;
  \item Explore the inner-structure of datasets.
  \end{itemize}

\section{Setting up the web interface}
\label{sec:configuring-sg-web}

  Before the web interface can be started, a few parameters have to be
  configured.  This is done through an XML file.  The following example
  displays all options, except for the authentication part, which is
  discussed separately in section \refer{sec:authentication}.

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  <fork>0</fork>
  <bind-address>127.0.0.1</bind-address>
  <port>8080</port>
  <home>/login</home>
  <developer-mode>0</developer-mode>
  <backtrace-on-error>0</backstrace-on-error>
  <system-connection>
    <uri>http://localhost:8890/sparql-auth</uri>
    <backend>virtuoso</backend>
    <username>dba</username>
    <password>dba</password>
  </system-connection>
  <authentication>
    <!-- Either LDAP settings, or local-user authentication -->
  </authentication>
</web-interface>
\end{lstlisting}

\subsection{To fork or not to fork}

  The \t{fork} property can be either \t{0} to keep the
  \program{sg-web} process in the foreground of your shell, or
  \t{1} to run the \program{sg-web} process as a daemon.

\subsection{Bind address and port or UNIX socket}

  To enable running multiple web services on single machine, \program{sg-web}
  can be configured to bind on an arbitrary address and an arbitrary port, or
  a UNIX domain socket.  For TCP sockets, both IPv4 and IPv6 addresses are
  supported.

  For UNIX domain sockets, use
  \t{<unix-socket>/path/to/socket-file</unix-socket>} instead of the
  \t{<bind-address>} and \t{<port>} configuration options.

\subsection{Default home page}

  When a user visits the \program{sg-web} instance, it defaults to showing
  the \t{/login} page.  When adding static pages to the \program{sg-web}
  instance, the \t{home} configuration option can be used to set a different
  default landing page.

  See section \refer{sec:static-pages} to read about how to add static pages
  to the \program{sg-web} instance.

\subsection{Developer mode}

  By default, changes to code in the \file{www/pages/} directory are
  applied when restarting \program{sg-web}.  When turning on
  `\t{developer-mode}', pages are reloaded when accessed, which makes
  interactive development of web pages easier.

\subsection{Logging and backtraces}
\label{sec:logging}

  The \program{sg-web} program keeps two logs: one for error reporting, and
  one for access and non-critical messages.  The log files can be specified
  as command-line options (\t{-e} for error reporting, \t{-d} for
  non-critical messages).

  Backtraces help the developer to get an idea about how something might've
  gone wrong.  It displays the location nearby the location in the source code
  that throwed an error.  Backstraces can get long and verbose.  Therefore we
  don't display them by default.  To enable reporting backtraces in the error
  log, set the `\t{backtrace-on-error}' option to \t{1}.

\subsection{System connection}

  The web interface stores its own information as RDF.  Therefore it needs
  a connection to an RDF store where it can write to the graphs described
  in table \ref{table:writable-graphs}.

  \hypersetup{urlcolor=black}
  \begin{table}[H]
    \begin{tabularx}{\textwidth}{l!{\VRule[-1pt]}X}
      \headrow
      \b{Graph} & \b{Reason}\\
      \evenrow
      \t{http://sparqling-genomics.org/sg-web/state}
      & In this graph, queries and projects are stored.\\
    \end{tabularx}
    \caption{\small Graphs that need to be writable for the web interface.}
    \label{table:writable-graphs}
  \end{table}
  \hypersetup{urlcolor=LinkGray}

\subsubsection{System connection example}

  To configure the \i{system connection}, two parameters need to be
  specified: \t{uri}, and \t{backend}.  Additionally, when the
  RDF store requires authentication for writing to it, a \t{username}
  and a \t{password} can be provided.

  The following example shows how to configure the \i{system connection}:

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  ...
  <system-connection>
    <uri>http://localhost:8890/sparql-auth</uri>
    <backend>virtuoso</backend>
    <username>dba</username>
    <password>dba</password>
  </system-connection>
</web-interface>
\end{lstlisting}

\subsection{Beacon support}
\label{sec:beacon}

  Beacon is an interface to create a ``global search engine for genetic
  mutations'' \citep{beacon-network}.  It achieves this by defining a
  standard that institutions must implement so that one search engine can
  query the implementations from each institution to find a specific
  genetic mutation.

  The web interface can function as a Beacon in the Beacon network
  \citep{beacon-network}.  The Beacon API uses a separate connection,
  similar to the \t{system-connection}, so that access can be
  controlled at the user level.

  The following example shows how to configure \i{Beacon} including the
  \i{Beacon connection}:

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  ...
  <beacon>
    <enabled>1</enabled>
    <organization>
      <id>SG</id>
      <name>SPARQLing-genomics Beacon service</name>
      <description>
        This Beacon service provides variant information for data hosted by
        this instance of the RDF store.
      </description>
      <address>Not provided</address>
      <welcome-url>https://www.sparqling-genomics.org</welcome-url>
      <contact-url>mailto:beacon@sparqling-genomics.org</contact-url>
      <logo-url>https://www.sparqling-genomics.org/static/images/logo.png</logo-url>
      <info>Not provided</info>
    </organization>
    <connection>
      <uri>http://localhost:8000</uri>
      <backend>virtuoso</backend>
      <username>beacon</username>
      <password>changeme</password>
    </connection>
  </beacon>
</web-interface>
\end{lstlisting}

  The implementation assumes the following conditions are met:
  \begin{itemize}
  \item Variant call data was imported using \t{vcf2rdf};
  \item The reference genome can be identified by the chromosome's NCBI
    identifier.
  \end{itemize}

\subsection{User management and authentication}
\label{sec:authentication}

  The \program{sg-web} assumes every user of the system has a username
  and a password to authenticate oneself. There are three ways to configure
  authentication.  For isolated deployments or environments, preconfigured
  accounts can be specified.  For organizational deployments, the web interface
  can be configured to use an LDAP server that supports version 3 of the LDAP
  protocol.  For public access deployments, ``log in with ORCID'' can be
  configured.

  Regardless of the authentication mechanism, there are a few reserved users
  that carry out a specific task within \program{sg-web}.  The table below
  describes the reserved users.

  \hypersetup{urlcolor=black}
  \begin{table}[H]
    \begin{tabularx}{\textwidth}{l!{\VRule[-1pt]}X}
      \headrow
      \b{Username} & \b{Task}\\
      \evenrow
      \t{beacon}
      & When the BEACON API (see section \refer{sec:beacon}) is enabled, and
      this user is added to a project, the data that is accessible by the
      project will be exposed to the BEACON network.\\
      \oddrow
      \t{portal}
      & This username has been reserved for an implementation of a data
      portal.\\
    \end{tabularx}
    \caption{\small Reserved users for \program{sg-web}.}
    \label{table:reserved-users}
  \end{table}
  \hypersetup{urlcolor=LinkGray}

\subsubsection{Local users configuration}

  The simplest form of authentication is the ``local-user configuration''.
  Configuring it involves providing a username and the SHA256 sum of a password
  for each account.  The following example shows how to configure
  ``local-user authentication'':

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  ...
  <authentication>
    <user>
      <username>user</username>
      <!-- The password field must contain the SHA256 sum of the
           plaintext password -->
      <password>9f86d08...0f00a08</password>
    </user>
    <user>
      <username>user2</username>
      <password>152f347...0b7a26a</password>
    </user>
  </authentication>
</web-interface>
\end{lstlisting}

\subsubsection{LDAP authentication example}

  To configure LDAP, four parameters must be specified: the URI to the LDAP
  service (1), optionally, an extra ``common name'' (2), optionally the
  ``organizational unit'' (3), and the ``domain'' (4).  The username is used
  as a ``common name''.

  \begin{sloppypar}
  Additionally, an alternative SSL certificate bundle can be configured with
  the parameters \t{ssl-certificate-directory} and
  \t{ssl-certificate-file}.
  \end{sloppypar}

  The following example shows how to configure LDAP authentication:

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  ...
  <authentication>
    <ldap>
      <uri>ldaps://example.local</uri>
      <common-name>AdditionalCN</common-name>
      <organizational-unit>People</organizational-unit>
      <domain>department.organization.tld</domain>
      <ssl-certificate-directory>/etc/ssl/certs</ssl-certificate-directory>
      <ssl-ca-certificate-file>
        /etc/ssl/certs/ca-certificates.crt
      </ssl-ca-certificate-file>
    </ldap>
  </authentication>
</web-interface>
\end{lstlisting}

\subsubsection{ORCID authentication example}

  To configure ``login with ORCID'', four parameters must be specified:
  the \t{client-id} (1) and \t{client-secret} (2) which are provided
  when registering your application at ORCID.  The \t{endpoint} (3) which
  should always be \t{https://orcid.org/oauth}, and the \t{redirect-uri}
  (4) that is used to redirect to when a user has logged into ORCID and
  authorized the login request.  The default path to redirect to should
  be \t{/login}.  Internally, the ORCID ID is used as identifier for a user.

  The following example shows how to configure ORCID authentication:

\begin{lstlisting}[language=XML]
<?xml version="1.0" encoding="utf-8"?>
<web-interface>
  ...
  <authentication>
    <orcid>
      <client-id>APP-XXXXXXXXXXXXXXXX</client-id>
      <client-secret>XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX</client-secret>
      <endpoint>https://orcid.org/oauth</endpoint>
      <redirect-uri>https://<your-public-uri>/login</redirect-uri>
    </orcid>
  </authentication>
</web-interface>
\end{lstlisting}

\subsection{Running the web interface}

  The web interface can be started using the \program{sg-web} command:

\begin{lstlisting}
sg-web --configuration-file=file.xml
\end{lstlisting}

  $\ldots{}$ where \t{file.xml} is a configuration file as
  discussed in section \refer{sec:configuring-sg-web}.

\pagebreak{}
\section{Managing multi-node setups with \program{sg-auth-manager}}

  When the knowledge graph grows beyond the capabilities of a single machine,
  one can consider scaling out to another.  The provided solution to manage
  multiple RDF stores while maintaining a single-point-of-entry for users
  involves running helper program called \program{sg-auth-manager} alongside
  a secondary RDF store.  This program communicates with \program{sg-web} to
  manage permissions, and handle data import and export.  Figure
  \ref{fig:sg-auth-manager} provides a schematic overview of the deployment
  model.

  \begin{figure}[H]
    \begin{center}
    \includegraphics[width=0.8\textwidth]{figures/sg-auth-manager-scaleout.pdf}
    \end{center}
    \caption{\textit{Illustrating a three-node setup.}}
    \label{fig:sg-auth-manager}
  \end{figure}

  %At least one instance of \program{sg-web} must run, so that the
  %\program{sg-auth-manager} can verify the validity of a request that was sent
  %to it.

  The \program{sg-auth-manager} program verifies the validity of the user's session
  token with the \program{sg-web} instance before acting as a proxy to the managed
  RDF store.

\subsection{Importing data into secondary stores}

  RDF importing is handled by an API call to the instance managed by
  \program{sg-auth-manager}.

  When \program{sg-web} receives a request to import data into an
  \program{sg-auth-manager}-managed RDF store, it passes the data directly
  to the \program{sg-auth-manager}, which in turn handles the importing of data
  into that store.

\subsection{How \program{sg-web} handles downtime of a \program{sg-auth-manager}}

  The \program{sg-auth-manager} registers itself with the pre-configured
  \program{sg-web} instance upon start.  In turn, the \program{sg-web} instance
  polls the availability of the \program{sg-auth-manager} instance every 10
  seconds.  When the \program{sg-auth-manager} instance is unavailable for 30
  consecutive seconds, \program{sg-web} removes the connection, and the
  \program{sg-auth-manager} instance must re-register itself.

\subsection{Federated querying of protected endpoints}

  The \t{sg-auth-manager} provides a mechanism to execute federated queries
  using protected endpoints by allowing the session token (see section
  \ref{sec:tokens}) as part of the URI in the \t{SERVICE} specification.

\pagebreak{}
\section{Testing a running instance of \t{sg-web}.}

  The test suite used to prevent functional regressions between releases is
  available in the \t{sg-web-test} program.  This program can be used to test
  whether all features from the API are working correctly after deployment.

  The program can perform two types of tests: \t{endpoint-test} and
  \t{sparql-parser-test}.

\subsection{Testing an endpoint}

  The \t{endpoint-test} interacts with an endpoint through its API.  To run
  this test, invoke \program{sg-web-test} as following:

\begin{lstlisting}
sg-web-test --endpoint https://www.sparqling-genomics.org --token TOKEN
\end{lstlisting}

  Replace \t{https://www.sparqling-genomics.org} with your endpoint's
  address, and \t{TOKEN} with an access token for your endpoint.

\subsection{Testing the SPARQL parser}

  To detect defects in the SPARQL parser, we use the \t{sparql-parser-test}
  to parse queries from \t{*.sparql} files from a directory.  To start this
  test, invoke \program{sg-web-test} as following:

\begin{lstlisting}
sg-web-test --sparql-parser /path/to/directory/with/queries
\end{lstlisting}

\pagebreak{}
\section{Using the web interface}
\label{sec:using-web-interface}

  Once the web interface is up and running, a logged-in user will land
  on the \i{Dashboard}.

  \begin{figure}[H]
    \begin{center}
      \includegraphics[width=1.0\textwidth]{figures/sg-web-dashboard.pdf}
    \end{center}
    \caption{On the \i{Dashboard}, tokens and connections can be
      configured.}
    \label{fig:web-dashboard}
  \end{figure}

  The dashboard provides access to two resources: \i{session tokens}
  and \i{RDF stores}.  Session tokens will be discussed in section
  \refer{sec:tokens}.  This leaves us with \i{RDF stores} for now.

\subsection{Configuring connections}
\label{sec:configure-connections}

  The web interface supports two types of connections; \i{user-specific
    connections} and \i{system-wide connections}.  Users can add the
  former using the \t{+} button.  Such a connection will only be visible
  to the user who added it.  On the contrary, \i{system-wide connections}
  are visible to all users.

  System-wide connections are therefore not configured by a user, but by a
  program called \program{sg-auth-manager}.  This program works as a middle-man
  between a private RDF store, and a running instance of \program{sg-web}.  It
  uses the user management from \program{sg-web} described in section
  \refer{sec:authentication} to allow users to access graphs assigned to one
  of their projects.

  An instance of \program{sg-auth-manager} registers itself with
  \program{sg-web} as a system-wide connection: making itself available to
  all users of the \program{sg-web} instance.

\subsection{Projects}
\label{sec:web-projects}

  The main way to access the knowledge graph, and to collaborate with other
  users is through a \i{project}.  Projects combine \i{users},
  \i{access to data} and \i{queries}.  The structure of a project
  attempts to capture general phases of analyzing data: \i{collect}
  $\rightarrow$ \i{structure} $\rightarrow$ \i{query} $\rightarrow$
  \i{report} $\rightarrow$ \i{automate}.

  \begin{figure}[H]
    \begin{center}
      \includegraphics[width=1.0\textwidth]{figures/sg-web-project-details.pdf}
    \end{center}
    \caption{The project overview page.}
    \label{fig:web-project-overview}
  \end{figure}

\subsubsection{Collecting data}

  Before data can be analyzed, it must be collected and stored.  Graphs are
  the primary place to store data.  Graphs are identified by a uniform resource
  identifier (URI).  So before data can be imported, a project must assign a
  graph.

  After assigning a graph to the project on the \i{Overview} page, the
  command to load a file can be generated in three steps.

  \begin{figure}[H]
    \begin{center}
      \includegraphics[width=1.0\textwidth]{figures/sg-web-import-rdf.pdf}
    \end{center}
    \caption{Importing RDF in three steps}
    \label{fig:web-import-rdf}
  \end{figure}

\subsection{Executing queries}

  After configuring at least one endpoint, it can be chosen on the \i{query}
  page to execute a query against it.

  \begin{figure}[H]
    \begin{center}
      \includegraphics[width=1.0\textwidth]{figures/sg-web-query.pdf}
    \end{center}
    \caption{The \i{query} page enables users to execute a query against a
      SPARQL endpoint.  The connections configured at the \i{connections} page
      can be chosen from the drop-down menu.}
    \label{fig:web-query}
  \end{figure}

\subsection{Query history}
\label{sec:query-history}

  When prototyping SPARQL queries, better known as ``SPARQLing around'', it's
  good to know that all queries that yielded a result are stored in the
  \i{query history}.  The history is shown on the \i{query} page below the
  query editor.  Each \i{project} has its own query history.

\subsection{Explore graphs with the Exploratory}

  Another utility aimed at SPARQLing around faster is the \i{exploratory}.

  \begin{figure}[H]
    \begin{center}
      \includegraphics[width=1.0\textwidth]{figures/sg-web-exploratory.pdf}
    \end{center}
    \caption{The \i{exploratory} page enables users to learn about the
      structure of the triplets in a graph.}
    \label{fig:web-exploratory}
  \end{figure}

  The exploratory uses a common pattern in RDF to help writing queries.  Its
  interface provides a four-step selection process to find \i{predicates}
  associated with an \t{rdf:type}.  The programs described in chapter
  \refer{chap:command-line} automatically add the \t{rdf:type} annotations.

\subsection{Connections and graphs}

  The first step in finding predicates involves choosing a connection
  (see section \refer{sec:configure-connections}).  The second step involves
  choosing a graph.  If the connection does not support the use of graphs,
  the journey ends here.

\subsection{Types}

  The third step looks for triplets that match the pattern \i{subject}
  $\rightarrow$ \t{rdf:type} $\rightarrow$ \i{type}.  All matches for
  \i{type} are displayed.  For data imported with \t{vcf2rdf} (see
  section \refer{sec:vcf2rdf}), this will display (among other types) the
  \t{VariantCall} type.

\subsection{Predicates}

  Staying with the \t{VariantCall} example;  All data properties extracted
  from a VCF file can be found under this type.  A predicate displayed in this
  column occurs in \i{at least} one triplet.  It not necessarily occurs in
  \i{every} triplet.  Especially when using \t{INFO} and \t{FORMAT}
  fields in a VCF file, we recommend using them in a query inside an
  \t{OPTIONAL} clause.

\pagebreak{}
\input{web-programming-interface}

\pagebreak{}
\input{web-extend}
